### DISCUSSION TITLE
Size-color sensitive object, how to train correctly?

### DISCUSSION THREAD

**Discussion by thanhminhmr** [ultralytics/yolov5#3499] [Q&A]
For context, I'm trying to create an OCR for simple text captcha. Currently I can generate simple text captchas with any specified size and color.

But since characters in captchas are sensitive to color and size (decoy characters have lower contrast than normal character, etc.) I don't know what setting I should train with. So my question is:

- Do the data loader _stretch_ the image to the specified size, or will it try to _fit_ the image in order to keep the original image ratio? (Captchas come with different size, but the character shape is the same, so ideally I should try to keep the original ratio of the image)
- Do the data augmenter change the color/ratio of the image, and how do I prevent that?

Note: This is what I mean when I say _stretch_ or _fit_ the image:
![image](https://user-images.githubusercontent.com/31412085/120992508-deba8480-c7ac-11eb-993c-38473321f590.png)

**Comment by glenn-jocher** [comment]
@thanhminhmr 👋 Hello! Thanks for asking about **image augmentation**. YOLOv5 🚀 applies online imagespace and colorspace augmentations in the trainloader (but not the testloader) to present a new and unique augmented [Mosaic](https://github.com/ultralytics/yolov5/blob/90b7895d652c3bd3d361b2d6e9aee900fd67f5f7/utils/datasets.py#L678-L732) (original image + 3 random images) each time an image is loaded for training. **Images are never presented twice in the same way**.

![YOLOv5 augmentation](https://user-images.githubusercontent.com/26833433/120995721-f3cfed00-c785-11eb-8ee2-b6ef2fa205e8.jpg)

The hyperparameters used to define these augmentations are in your hyperparameter file (default `data/hyp.scratch.yaml`) defined when training:
```
python train.py --hyp hyp.scratch.yaml
```
https://github.com/ultralytics/yolov5/blob/90b7895d652c3bd3d361b2d6e9aee900fd67f5f7/data/hyp.scratch.yaml#L1-L33

You can view the effect of your augmentation policy in your train_batch*.jpg images once training starts. These images will be in your train logging directory, typically `yolov5/runs/train/exp`:

`train_batch0.jpg` shows train batch 0 mosaics and labels:  
> <img src="https://user-images.githubusercontent.com/26833433/83667642-90fcb200-a583-11ea-8fa3-338bbf7da194.jpeg" width="700">


Good luck and let us know if you have any other questions!
